### PR TITLE
Reshuffle and reduce test cases

### LINKED ISSUE #816: [PR Testing] Get rid of test redundancy

### LINKED ISSUE #877: [COMGR] Code quality: reference binding to null pointer of type 'char'
The GetBytes function, takes a reference to the first element of the zero-size vector, which results in an error because the element does not exist

Should be checked in GetString, before calling GetBytes and if (GetSize() ==0)  return empty string {} ?

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/00d57548c25399186eb201eb2e0e6f7480915c48/src/comgr.cpp#L532-L554

To reproduce:
```
CXX='/opt/rocm/llvm/bin/clang++' CXXFLAGS='-Werror' cmake -DMIOPEN_TEST_FLAGS=' --disable-verification-cache  --verbose'  -DCMAKE_BUILD_TYPE=debug -DCMAKE_CXX_FLAGS_DEBUG='-g -fno-omit-frame-pointer -fsanitize=undefined -fno-sanitize-recover=undefined'  -DBUILD_DEV=On -DMIOPEN_GPU_SYNC=Off -DMIOPEN_USE_COMGR=On ../
```

Error log:

```
MIOpen(HIP): Info [FindSolutionImpl] ConvOclDirectFwd1x1
MIOpen(HIP): Info [FindSolutionImpl] Perf Db: record not found for: ConvOclDirectFwd1x1
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.1.9.0, USE_HIP_PCH: 1
/usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_vector.h:798:9: runtime error: reference binding to null pointer of type 'char'
    #0 0x7fffedb7a280  (/home/kamil/vf/MLOpen-build/lib/libMIOpen.so.1+0x309d280)
    #1 0x7fffedb984df  (/home/kamil/vf/MLOpen-build/lib/libMIOpen.so.1+0x30bb4df)
    #2 0x7fffedbac8ad  (/home/kamil/vf/MLOpen-build/lib/libMIOpen.so.1+0x30cf8ad)
    #3 0x7fffedb8b9b3  (/home/kamil/vf/MLOpen-build/lib/libMIOpen.so.1+0x30ae9b3)
```

From gdb:
```
MIOpen(HIP): Info [FindSolutionImpl] Perf Db: record not found for: ConvOclDirectFwd1x1
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.1.9.0, USE_HIP_PCH: 1

Thread 1 "test_conv2d" received signal SIGILL, Illegal instruction.
0x00007fffeea07eca in std::vector<char, std::allocator<char> >::operator[] (this=0x7fffffff1c28, __n=0)
    at /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_vector.h:798
798             return *(this->_M_impl._M_start + __n);
(gdb) backtrace full 3
#0  0x00007fffeea07eca in std::vector<char, std::allocator<char> >::operator[] (this=0x7fffffff1c28, __n=0)
    at /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_vector.h:798
No locals.
#1  0x00007fffeea1e9c4 in miopen::comgr::Data::GetBytes (this=0x7fffffff1cf0, bytes=std::vector of length 0, capacity 0)
    at /home/kamil/vf/MLOpen/src/comgr.cpp:545
        status = 32767
        sz = 0
#2  0x00007fffeea2d7bd in miopen::comgr::Data::GetString[abi:cxx11]() const (this=0x7fffffff1cf0) at /home/kamil/vf/MLOpen/src/comgr.cpp:551
        bytes = std::vector of length 0, capacity 0
        sz = 0
```

### PR DESCRIPTION
**Pull request by daniellowell**
- Limit the number of combinations for a single dimension for pooling and convolution tests
- Resolves #816
- Resolves #877

features:
- test generator, now it includes a batch greater than 1 and able to variate count of tests using **--limit** `MIOPEN_TEST_LIMIT` for cmake
- varies improvements in CMakeLists. Fixed LONG_TESTS, added information about skipped tests.
- standartized Jenkinsfile
- reshuffled test stage sequence

active WAs:
- #898
- #936

### CODE REVIEW

File: test/conv_common.hpp [hunk 1]
```diff
@@ -1486,13 +1534,48 @@ struct conv_driver : test_driver
...
+                input_channels,
+                spatial_dim_elements.at(0),
+                spatial_dim_elements.at(1),
+                spatial_dim_elements.at(
+                    2)}.generate(tensor_elem_gen_integer{17});
+        }
+        else
+        {
+            MIOPEN_LOG_E("dimension is wrong!");
+        }
```
Comment (pfultz2) [inline test/conv_common.hpp:1576]: This breaks passing in the configs with `--input` and `--weights`, which means the deepbench test_conv_extra tests are not working correctly. Also, it makes it harder to pass the config in from the log. 

It also makes it harder to control what cases we want to run.

Comment (daniellowell) [reply test/conv_common.hpp:1576]: I plan on updating to use --input and --weights as overrides to the explicit batch-channel-filter usage. I think these only make it harder if you don't like putting in more than 2 flags, other than that it is pretty clear what is happening. In any case they users should be able to fall back to the original inputs.

Comment (pfultz2) [reply test/conv_common.hpp:1576]: Yea, but you still have no way of avoiding too many cross product combinations or duplicating config we are already testing. 

You could use a for loop:

```cpp
std::vector<std::vector<std::size_t>> get_inputs()
{
    std::vector<std::vector<std::size_t>> result;
    for(auto batch_size:get_batch_sizes())
        for(auto batch_size:get_input_channels())
            for(auto spatial_dim_elements:get_spatial_dims())
            {
                std::vector<std::size_t> dim = {batch_size, input_channels};
                dim.insert(dim.end(), spatial_dim_elements.begin(), spatial_dim_elements.end());
                result.push_back(dim);
            }
    return result;
}
```

Although, I think explicitly listing them is better so we can better coverage with less tests.

**Comment by atamazov** [comment]
@daniellowell
> No half tests on OpenCL backend. 

Are we officially phasing out support for FP16 with OCL? If not, then I recommend keeping minimal testing at least.

**Comment by daniellowell** [comment]
We don’t have support for fp16 on OpenCL. MIOpenGemm does not support it.

**Comment by JehandadKhan** [comment]
@daniellowell I suggest we merge this with what we already have and only open another PR for future changes.

**Comment by daniellowell** [comment]
Problem is I keep finding bugs in either the stack, or our library components. This requires me to workaround it and push the devs to fix the issue. I think this round of testing will succeed....I think.

**Review by DrizztDoUrden — DISMISSED** [review]

**Comment by atamazov** [comment]
@DrizztDoUrden Please review (I should too)

**Comment by atamazov** [comment]
@daniellowell Failed several times in a row during build at `Long Tests I`/`GCC Release conv2d`

**Comment by atamazov** [comment]
@shurale-nkn `removeHcc` partially depends on this. The branching commit is d06c645006bb49737aec7c280d8c0c6416420f67 `Missing step.`

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. You can continue work as usual. But please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

**Comment by junliume** [comment]
@atamazov are we maintaining this work in progress PR please, maybe by rebasing from time to time? As time went by it might be difficult to merge if we keep it as is, if that's the ultimate intention. Thanks!

**Comment by atamazov** [comment]
@junliume This is developer's responsibility. AFAICS the recent develop was merged into this branch 15 days ago. Only Jenkinsfile is having merge conflicts, which should be easy to resolve.

**Comment by atamazov** [comment]
@shurale-nkn **_I am going to merge this PR next_**. Please merge recent develop. I would like to make sure that recent updates in develop do not break anything.

**Comment by atamazov** [comment]
well, let me do that myself, this would be faster.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -861,8 +925,56 @@ COMMAND	$<TARGET_FILE:test_conv2d>	--verbose	--input	32	64	27	27	--weights	192	6
...
+if(${MIOPEN_USE_MIOPENGEMM} AND (${MIOPEN_hip_VERSION_MAJOR} EQUAL 3) AND (${MIOPEN_hip_VERSION_MINOR} EQUAL 7))
+    add_custom_test(test_conv3d_extra  SKIP_UNLESS_ALL
+    COMMAND MIOPEN_LOG_LEVEL=6 $<TARGET_FILE:test_conv3d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 50 50 50 --weights 32 16 5 5 5 --pads_strides_dilations 0 0 0 1 1 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+    COMMAND MIOPEN_LOG_LEVEL=6 $<TARGET_FILE:test_conv3d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 50 50 50 --weights 32 16 5 5 5 --pads_strides_dilations 0 0 0 2 2 2 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+    COMMAND MIOPEN_LOG_LEVEL=6 $<TARGET_FILE:test_conv3d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 50 50 50 --weights 32 16 5 5 5 --pads_strides_dilations 2 2 2 1 1 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+    COMMAND MIOPEN_LOG_LEVEL=6 $<TARGET_FILE:test_conv3d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 50 50 50 --weights 32 16 5 5 5 --pads_strides_dilations 0 0 0 1 1 1 2 2 2 ${MIOPEN_TEST_FLAGS_ARGS}
+    #COMMAND $<TARGET_FILE:test_conv3d> ${MIOPEN_TEST_FLOAT_ARG} --input 1 16 4 161 700 --weights 16 16 3 11 11 --pads_strides_dilations 1 1 1 1 1 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+
+    #ROCM3.7 compiler problems
```
Comment (atamazov) [inline test/CMakeLists.txt:955]: This must be uncommented.

**Review by atamazov — APPROVED** [review]
Fix separately please

File: Jenkinsfile [hunk 3]
```diff
@@ -270,778 +314,414 @@ pipeline {
...
-                    environment{
-                        cmd = """
-                            ulimit -c unlimited
-                            cd build
-                            CXX=/opt/rocm/llvm/bin/clang++ cmake -DMIOPEN_USE_MLIR=On -DMIOPEN_TEST_GFX908=On -DBUILD_DEV=On -DCMAKE_BUILD_TYPE=release -DMIOPEN_TEST_FLAGS='--verbose --disable-verification-cache' ..
-                            CTEST_PARALLEL_LEVEL=4 MIOPEN_LOG_LEVEL=5 make -j\$(nproc) check
-                        """
-                    }
+                stage('Fp16 Hip MLIR Xdlops') {
+                    agent{ label rocmnode("vega") }
```
Comment (atamazov) [inline Jenkinsfile:446]: This should be tested on gfx908.

File: Jenkinsfile [hunk 4]
```diff
@@ -270,778 +314,414 @@ pipeline {
...
-                    environment{
-                        cmd = """
-                            ulimit -c unlimited
-                            cd build
-                            CXX=/opt/rocm/llvm/bin/clang++ cmake  -DMIOPEN_TEST_HALF=On -DMIOPEN_USE_MLIR=On -DMIOPEN_TEST_GFX908=On -DBUILD_DEV=On -DCMAKE_BUILD_TYPE=release -DMIOPEN_TEST_FLAGS='--verbose --disable-verification-cache' ..
-                            CTEST_PARALLEL_LEVEL=4 MIOPEN_LOG_LEVEL=5 make -j\$(nproc) check
-                        """
-                    }
+                stage('Fp32 Hip MLIR Xdlops') {
+                    agent{ label rocmnode("vega") }
```
Comment (atamazov) [inline Jenkinsfile:452]: Ditto

**Review by atamazov — COMMENTED** [review]
MLIR is not tested on gfx908.

File: Jenkinsfile [hunk 5]
```diff
@@ -270,778 +314,414 @@ pipeline {
...
-                            cd build
-                            CXX=/opt/rocm/llvm/bin/clang++ cmake -DMIOPEN_TEST_GFX908=On -DMIOPEN_TEST_ALL=On -DBUILD_DEV=On -DCMAKE_BUILD_TYPE=release -DMIOPEN_TEST_FLAGS='--verbose --disable-verification-cache' ..
-                            MIOPEN_LOG_LEVEL=5 CTEST_PARALLEL_LEVEL=4 MIOPEN_CONV_PRECISE_ROCBLAS_TIMING=0 make -j\$(nproc) check
-                        """
-                    }
                     steps{
-                        script{
-                            runDockerJob(cmd: cmd, gpu_arch: "gfx908")
-                        }
+                        buildHipClangJobAndReboot(setup_flags: Full_test_limit + gfx908_test, build_env: WORKAROUND_iGemm_936, gpu_arch: "gfx908")
```
Comment (atamazov) [inline Jenkinsfile:572]: @shurale-nkn WORKAROUND_iGemm_936 is for HALF and not needed for FP32 stages. I will fix this soon.

File: test/conv_common.hpp [hunk 6]
```diff
@@ -1661,6 +1730,69 @@ struct conv_driver : test_driver
...
+            input = tensor<T>{
+                batch_size,
+                input_channels,
+                spatial_dim_elements.at(0),
+                spatial_dim_elements.at(1),
+                spatial_dim_elements.at(
+                    2)}.generate(tensor_elem_gen_integer{17});
+        }
+
+        if(!weight_tensor_dims.empty())
```
Comment (atamazov) [inline test/conv_common.hpp:1771]: ⚓

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -861,8 +925,56 @@ COMMAND	$<TARGET_FILE:test_conv2d>	--verbose	--input	32	64	27	27	--weights	192	6
...
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 3072 3072 --weights 32 16 3 3 --pads_strides_dilations 0 0 2 2 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 2 16 3072 3072 --weights 32 16 3 3 --pads_strides_dilations 2 2 2 2 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 128 320 1 7 --weights 256 320 1 1 --pads_strides_dilations 0 0 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 128 1024 1 7 --weights 2048 1024 1 1 --pads_strides_dilations 1 1 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 352 192 7 1 --weights 320 192 1 1 --pads_strides_dilations 0 0 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_conv2d> ${MIOPEN_TEST_FLOAT_ARG} --input 352 16 7 1 --weights 32 16 1 1 --pads_strides_dilations 2 2 1 1 1 1 ${MIOPEN_TEST_FLAGS_ARGS}
 )
 
+if (0) #disabled too many errors
```
Comment (atamazov) [inline test/CMakeLists.txt:945]: ⚓

**Pull request merged**